File: Exercise_1.py
# ## Problem1
# Subsets (https://leetcode.com/problems/subsets/)
# // Time Complexity : O(2^n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No


# // Your code here along with comments explaining your approach

# S1: RQ
# S2: Constraints
# S3: Test Cases
# S4: Solution without Code
# 0-1 solution
# S5: Solution with Code


# For loop based Recursion using Backtracking
class Solution:
    def subsets(self, nums):
        # result list
        self.result = []
        self.helper(nums, 0, [])
        return self.result

    def helper(self, nums, pivot, path):
        self.result.append(list(path))
        for i in range(pivot, len(nums)):
            # action
            path.append(nums[i])
            # recurse
            self.helper(nums, i + 1, path)
            # backtrack
            path.pop()

# // Space Complexity : O(n)
# Bactracking solution
# class Solution:
#     def subsets(self, nums):
#         # result list
#         self.result = []
#         self.helper(nums, 0, [])
#         return self.result

#     def helper(self, nums, i, path):
#         # base
#         if i == len(nums):
#             # creating a deep copy of path when adding to result so it doesnt get backtracked
#             self.result.append(list(path))
#             return
#         # logic
#         # not choose - case 0
#         self.helper(nums, i + 1, path)
#         # choose - case 1
#         # action
#         path.append(nums[i])
#         # recurse
#         self.helper(nums, i + 1, path)
#         # backtrack
#         path.pop()

# // Space Complexity : O(2^n)
# 0-1 solution
# class Solution:
#     def subsets(self, nums):
#         # result list
#         self.result = []
#         self.helper(nums, 0, [])
#         return self.result

#     def helper(self, nums, i, path):
#         # base
#         if i == len(nums):
#             appending path to final result
#             self.result.append(path)
#             return

#         # logic
#         # not choose - case 0
#         self.helper(nums, i + 1, list(path))
#         # choose - case 1
#         path.append(nums[i])
#         self.helper(nums, i + 1, list(path))


nums = [1, 2, 3]
sol = Solution()
print(sol.subsets(nums))

File: Exercise_1.cpp
/* 
    Time Complexity                              :  O(N!) since at every step the choice reduces by 1 
                                                    as no duplicates are allowed
    Space Complexity                             :  O(N) - size of the Nums array since the recursion stack
                                                    will go maximum to the number of elements in the NUms array.
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

// https://leetcode.com/problems/subsets/

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        sz = nums.size();
        vector<int> lv;
        lv.clear();
        // subs3(nums, 0, lv);
        // return ans;
        
        return subs4(nums);
    }
    
private:
    vector<vector<int>> ans;
    int sz;
    
    // DP
    vector<vector<int>> subs4(vector<int>& nums) {
        
        vector<vector<vector<int>>> dp(sz+1);
        dp[0] = vector<vector<int>>(1,vector<int>());
        
        for(int i=1;i<=sz;i++) {
            vector<vector<int>> vec = dp[i-1];
            int n = nums[i-1];
            dp[i] = dp[i-1];
            for(auto v : vec) {
                v.push_back(n);
                dp[i].push_back(v);
                }
            }
        return dp[sz];
    }
    
    // backtracking
    void subs(vector<int>& nums, int ci, vector<int>& lv) {
        ans.push_back(lv);
        
        for(int i=ci;i<sz;i++) {
            lv.push_back(nums[i]);
            subs(nums,i+1,lv);
            lv.pop_back();
        }
    }
    
    // recursion
    void subs2(vector<int>& nums, int ci, vector<int> lv) {
        if(ci == sz) {
            ans.push_back(lv);
            return;
        }
        
        subs(nums, ci+1, lv);
        lv.push_back(nums[ci]);
        subs(nums,ci+1,lv);
    }
    
    // for loop based recursion
    void subs3(vector<int>& nums, int ci, vector<int> lv) {
        ans.push_back(lv);
        for(int i=ci; i<sz; i++) {
            vector<int> tv(begin(lv),end(lv));
            tv.push_back(nums[i]);
            subs(nums, i+1, tv);
        }
    }
    
    
    
};